utils/zerovec/src/ule/niche.rs

```diff
@@ -0,0 +1,184 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::mem::size_of;
+use core::ops::DerefMut;
+use core::{marker::Copy, ops::Deref};
+
+use super::{AsULE, ULE};
+
+/// The types implementing this trait guarantee that [`NicheBytes::INVALID_BIT_PATTERN`]
+/// can never occur as a valid byte representation of the type.
+/// The implementors guarantee that N == core::mem::sizeo_of::<Self>()
+pub trait NicheBytes<const N: usize> {
+    const INVALID_BIT_PATTERN: [u8; N];
+}
+
+/// [`ULE`] type for [`NichedOption<U,N>`] where U implements [`NicheBytes`].
+/// The invalid bit pattern is used as the niche.
+///
+/// This uses 1 byte less than [`crate::ule::OptionULE<U>`] to represent [`NichedOption<U,N>`]
+/// which Derefs to [`Option<U>`].
+/// The implementors guarantee that N == core::mem::sizeo_of::<Self>()
+///
+/// # Example
+///
+/// ```rust
+/// use zerovec::ZeroVec;
+/// use zerovec::ule::NichedOption;
+/// use core::num::NonZeroU8;
+///
+/// let bytes = &[0x00, 0x01, 0x02, 0x00 ];
+/// let zv_no: ZeroVec<NichedOption<NonZeroU8, 1>> =
+///         ZeroVec::parse_byte_slice(bytes).expect("Unable to parse as NichedOption.");
+///
+/// assert_eq!(zv_no.get(0).as_deref(), Some(&None));
+/// assert_eq!(zv_no.get(1).as_deref(), Some(&NonZeroU8::new(1)));
+/// assert_eq!(zv_no.get(2).as_deref(), Some(&NonZeroU8::new(2)));
+/// assert_eq!(zv_no.get(3).as_deref(), Some(&None));
+/// ```
+// Invariants:
+// The union stores [`NicheBytes::INVALID_BIT_PATTERN`] when None.
+// Any other bit pattern is a valid.
+#[repr(packed)]
```

Comment:
Question: does `#[repr(packed)]` do anything on single-value unions? I guess it's harmless

Reply:
Changed to `[repr(c)]`